### ISSUE TITLE
Feature Request for Criterions: Add Wasserstein loss for W-GAN

### ISSUE THREAD

**Issue by ustcjerry** [torch/nn#1190]
As described in this [paper](http://arxiv.org/abs/1701.07875) and the [code](https://github.com/martinarjovsky/WassersteinGAN) in PyTorch. 
We need to define a new loss which is not like  binary cross-entropy like `BCECriterion` but just like this:

binary cross-entropy like `BCECriterion` 
```
D_loss = -(torch.mean(torch.log(D_real) + torch.log(1- D_fake)))
```
===============
Wasserstein loss is described below:

![00](https://cloud.githubusercontent.com/assets/11849381/24755454/dd7aa1b4-1b0c-11e7-8f0b-bf873d3b88c5.png)
where :
```
D_loss = -(torch.mean(D_real) - torch.mean(D_fake))
```
with weight clipping:
```
for p in D.parameters():
    p.data.clamp_(-0.01, 0.01)
```

The details are in this [blog](http://wiseodd.github.io/techblog/2017/02/04/wasserstein-gan/), and the corresponding PyTorch [code](https://github.com/wiseodd/generative-models/blob/master/GAN/wasserstein_gan/wgan_pytorch.py)
